R/CompDbSource.R

```diff
@@ -221,196 +293,258 @@ setClass("MetFragParam",
-				peaks <- peaks[-c(1:nrow(peaks)),]
-				## ...the other option is to ignor these calls - leaving the input as-is and returning empty results:
-				# out[[i]] = list(
-					# mfResults = NULL,
-					# mfServerStatus = "NA"
-				# )
-				# next()				
-			}
-			
-			peakListString <- paste(apply(peaks, 1, function (x){ paste(x[1],x[2],sep = '_') }),collapse = ';')
-
-			mz <- spectraData(query[i])$precursorMz
-
-			chkAdducts <- validAdducts(query,param)	
-			
-			if (!all(chkAdducts)) { 
-				stop("Some spectra precursor adducts are not legal MetFrag adduct types: Use function 'validAdducts(MetFragParam())' to print a list of valid adduct names. 'matchSpectra' stopping.") 
-			}
-							
-			adduct <- spectraData(query)$precursorAdduct[i]
-			## All REST calls will be done inside this loop here:
-			out[[i]] <- lapply(1:length(adduct),function (i) {
-				neutralPrecursorMass <- MetaboCoreUtils::mz2mass(mz, adduct[i])
-				restQuery <- sapply(slotNames(param)[-which(slotNames(param) %in% c("validAdducts","THRESHFUN"))], function(x) { 
-					slot(param,x)
-				})
-				restQuery <- c(
-					restQuery,	
-					"neutralprecursormass" = as.character(neutralPrecursorMass),
-					"precursoriontype" = as.character(adduct),
-					"peakliststring" = peakListString
+            qi <- query[i]
+			peaks <- peaksData(qi)[[1L]]
+			if (nrow(peaks) < param@minFragCount) {
+                out[[i]] <- data.frame(
+                    server_status = rep(NA_character_, length(adduct)),
+                    query_idx = rep(i, length(adduct)),
+                    adduct = adduct)
+                pb$tick()
+				next()
+			} else
+                peakListString <- .peaks_to_metfrag_string(peaks)
+			pmz <- qi$precursorMz
+			adduct <- unlist(qi$precursorAdduct)
+			out[[i]] <- do.call(rbindFill, lapply(adduct, function(a) {
+				query <- c(
+					restQuery,
+					neutralprecursormass = as.character(mz2mass(pmz, a)),
+					precursoriontype = a,
+					peakliststring = peakListString
 				)
-				
-				mfRequest <- POST(url = target@url,config = target@config,body = restQuery,encode = "json",verbose())
-						
-				## get the http call status, plus results(if exist): 
-				callStatus <- .chk_rest_response (mfRequest)
+				mfRequest <- POST(url = target@url, config = target@config,
+                                  body = query, encode = "json", handle)
+				callStatus <- .chk_rest_response(mfRequest)
 				if (callStatus$status == "SUCCESS" ) {
-					mfCandidates <- read.csv(text = content(GET(callStatus$rest_url))) 
-				} else {				
-					mfCandidates = NULL
-					## Warning msg?
-				}
-				## return a list of anntation results table and server staus code
-				list(
-					mfResults = mfCandidates,
-					mfServerStatus = callStatus$status
-				)
-			})
-			names(out[[i]]) <- adduct
-			Sys.sleep(1)
+					res <- read.csv(text = content(GET(callStatus$rest_url)))
+                    res$server_status <- callStatus$status
+                    ## filter based on score?
```

Comment:
shouldn't we filter somehow the results? e.g. if the `"score"` is higher than a certain value (defined by `@THRESHFUN`?

Reply:
I need to initially let all predictions be available for testing and performance evaluations - but also later - should it not be up to the user to filter results by any arbitrary criteria?